test(RewardsStreamerMP): ensure bonusMP and maxMP are decreased

0x-r4bbit opened the pull request:
correctly at unstake

This commit adds some tests to check that, if a user (partically) unstakes their funds, their initial and bonus MP get decreased proportionally, as well as their max mp.

Closes #46

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [ ] Added natspec comments?
- [x] Ran `pnpm adorno`?
- [ ] Ran `pnpm verify`?

0x-r4bbit commented:
This should also address #43

0x-r4bbit commented on test/RewardsStreamerMP.t.sol:
```diff
@@ -1102,6 +1102,89 @@ contract UnstakeTest is StakeTest {
+                rewardIndex: 0,
+                accountedRewards: 0
+            })
+        );
```
This passes, but I was wondering something.

In this scenario, I've staked for a year, and I locked up my funds for 3 months.
For locking up my funds, I've received a higher initial MP and my max MP has increase as well.
Cool.

Now a year has passed, I'm unstaking half of my funds, which means I should also loose half of my **initial** MP.
But should I also loose half of the bonus MP for which I have locked up my funds for 3 months initially?

I guess what I'm unsure about is, should I really loose the bonus MP for which I've done my work?
I might no longer stake the 10e18 tokens, but I did for 3 months, for which I'm entitled to that bonus.
Right now that is taken away from me when I partially unstake.

I wonder if we should keep, at least as long as the user is staked in the system.

Or maybe this is total nonsense.

Curious about your thoughts @gravityblast @3esmit

0x-r4bbit replied:
Okay, actually on a second thought I don't think this is a good idea.

If we'd keep MPs for the lock up that was done, we'd have to keep track of all lock ups per stake to make sure this checks out.

3esmit commented:
The test seem biased. It dosent seem to check for proportional reduce in MPs. But I might not be totally familiar with your test script. I'll submit a commit with a test I think is correct for this behavior. I'll also make the same on Stake Manager, as there it could also not be correct as we changed the logic a lot and dont have all proper tests for unstaking (we focused a lot on process account, epoch and estimation of MPs)

3esmit commented:
It looks biased because the tests are against hardcoded values instead of using calculations based on the inputs.

3esmit commented:
I reviewed the math from scratch, and determined that it indeed would return the correct values for when $A_{balance}$ increases, accrues, and later decreases by unstake.

RewardsStreamerMP uses the following formula:

$$
MP_{total_2} = MP_{total_1} - \frac{MP_{total_1} \times A_{reduced}}{A_{balance_1}}
$$


I found a simplification for this formula

$$
MP_{total_2} = A_{balance_1} \times \frac{MP_{total_1}}{A_{balance_1}}
$$

However, **for increasing** $A_{balance}$ (restake) **this will not work**, because $MP_{accrued}$ (and consequently $MP_{total}$) **is not retroactively increased**, unless the calculation is done incrementally (i.e., based on the balance during each respective period).

---

Also, I didnt understand yet why the $MP_{bonus}$ is being calculated as it is.
Currently RewardsStreamerMP uses:

$$
MP_{bonus} = A_{balance} * \frac{\left(\frac{t_{locked} \times M_{MAX} \times {SCALE_{FACTOR}}} {t_{MAX}} \right)}{SCALE_{FACTOR}}
$$

Why not use the normal formula?

$$
MP_{\text{bonus}} = A_{balance} + \left( \frac{A_{balance} \times t_{locked}}{t_{year}} \right)
$$


Maybe that should be another issue opened?

---

About the commit I submitted, I recognize it is currently wrong, I am still having to familiarize with the formulas on this contract and the difference in use storing $MP_{max}$ instead of $MP_{bonus}$, which I believe is a good improvement as it reduces calculations (I'll do that in StakeManager as well)

0x-r4bbit commented:
> About the commit I submitted, I recognize it is currently wrong [...] I'll fix them and create a test for the scenario with stake + time passes + unstake + time passes + stake + time passes + unstake + (...) to test if this is actually a bug, and if it is, how impactful, and what can be done to fix or mitigate it.

Awesome.
Given that this PR otherwise provides the correct tests and the math seems right, would you mind adding your test refinemint in a separate PR? 
That would unblock this one so we can get it merged.

0x-r4bbit commented:
> Maybe that should be another issue opened?

Will create an issue to discuss this.

0x-r4bbit commented:
> RewardsStreamerMP uses the following formula:
>
> $$
> MP_{total_2} = MP_{total_1} - \frac{MP_{total_1} \times A_{reduced}}{A_{balance_1}}
> $$
>
>
> I found a simplification for this formula
> 
> $$
> MP_{total_2} = A_{balance_1} \times \frac{MP_{total_1}}{A_{balance_1}}
> $$
>
> However, **for increasing** $A_{balance}$ (restake) **this will not work**, because $MP_{accrued}$ (and consequently > $MP_{total}$) **is not retroactively increased**, unless the calculation is done incrementally (i.e., based on the balance > during each respective period).
>
>To solve that we would need to store $MP_{bonus}$ (or extract it from $$MP_{max} - MP_{MAX}(A_{balance})$$), so the >$MP_{accrued}$ can be extracted from $MP_{total}$ and the $MP_{bonus_2}$ can be recalculated from $t_{locked}$ (pro-rated estimated from $A_{balance}$ and $MP_{bonus}$), while $MP_{accrued}$ is proportionally reduced.
Another option would be to save timestamp of last balance change and use it to determine all other values.

@3esmit so you're saying, we either use the formula you've suggest, which however would require more changes to properly update the contract's state, or we just stick with the formula that's currently implemented?

3esmit commented:
The simplification changes nothing besides less mathematical operations. To fix the issue we would have to implement a way of recalculating bonusmp so it can be increased separately

0x-r4bbit commented on test/RewardsStreamerMP.t.sol:
```diff
@@ -1224,21 +1224,24 @@ contract UnstakeTest is StakeTest {
             totalStaked[stage] = amountStaked;
-            predictedBonusMP[stage] = _calculateBonusMP(totalStaked[stage], secondsLocked);
+            predictedBonusMP[stage] = totalStaked[stage] + _calculateBonusMP(totalStaked[stage], secondsLocked);
+            predictedTotalMaxMP[stage] = (totalStaked[stage] * streamer.MAX_MULTIPLIER()) + _calculateBonusMP(totalStaked[stage], secondsLocked);
```
This is incorrect. `maxMP` should not include the `bonusMP`.
Had a discussion with @gravityblast and @mart1n-xyz about this just a few days ago.

Generally, your `bonusMP` should not affect your `maxMP`. Only your stake amount does.

3esmit replied:
MaxMP is the MAX_MULTIPLIER of _amount deposited. However, when we have bonusMP, the MaxMP would be shifted by the amount of bonus generated. Why this would be incorrect?

0x-r4bbit replied:
As mentioned above, I was unsure about it as well and had this checked once and for all (here's the discussion: https://discord.com/channels/864066763682218004/1282974895122419742/1294302914499837963)

0x-r4bbit replied:
As discussed offline, there was a misunderstanding (once again) about this.
I need to update this PR accordingly.

0x-r4bbit commented:
@3esmit Thanks again for providing your additional test.

I've rebased all changes on top the latest changes to the `user.maxMP` and `totalMaxMP`.
I've adjusted one line in the test you've provided, according to the changes.

The test now passes.

You can see the commit I've added to check.

Once you and @gravityblast give green light on this, I'm going to squash the last three commits into your commit that adds the additional test before this gets merged.

0x-r4bbit commented:
I've squashed [1f42c71](https://github.com/vacp2p/staking-reward-streamer/pull/48/commits/1f42c71f1aeed5893be566d11f4e320df9f7d2be) in to [bd9c683](https://github.com/vacp2p/staking-reward-streamer/pull/48/commits/bd9c6830167f7fca2e6a6714c1d8b5cd74c19651) and pushed again. Will merge once green
